File: scripts/02_makeCSVTable_csmatch.R
# Generate a CSV file based on the ScoreAcc files generated before
# Input:
# 1. Output file
# 2. Folder ID that needs to be added
# 3. expParents


library("optparse")
library("jsonlite")

option_list = list(
  make_option(c("-o", "--outFile"), type="character", default="intermediate_modified.csv", help="Output file name [default= %default]", metavar="character"),
  make_option(c("-f", "--folID"), type="character", default=NULL, help="Plate ID to be added in the CSV file", metavar="character"),
  make_option(c("-e", "--expParents"), type="character", default=".x.", help="expected parents, ex: 6090x9416", metavar="character")
);
opt = parse_args(OptionParser(option_list=option_list));

outFile = opt$outFile
id = opt$folID
expParents = opt$expParents

if(file.exists(outFile)){
  file.remove(outFile)
}

wind.filelist <- list.files("./", pattern = "[.]windowscore.txt$")
Names <- character()
plate <- character()
GivParents <- character()
FouParents1 <- character()
FouParents2 <- character()
TopHit <- character()
TopScore <- numeric()
NextHit <- character()
FracScore <- numeric()
LikeLihoodTopHit <- numeric()
LLRNextHit <- numeric()
TopHitsNumber <- numeric()
ScoreP1 <- numeric()
ScoreP2 <- numeric()
NumWindows <- numeric()
CountP1 <- numeric()
CountP2 <- numeric()
SNPscalled <- numeric()
PerHeterozygosity <- numeric()
Overlap <- numeric()
AmbiWindows <- character()
AmbiWindowCount <- character()
TopHitsNumber <- numeric()
for (echscore in wind.filelist){
  filename <- sub("[.]windowscore[.]txt", "", echscore)
  name <- sub("[.]csmatch", "", sub("[.]filter", "", sub("[.]snpmatch", "", filename)))
  likeLiwind <- try(read.table(echscore, header = F, as.is = TRUE))
  ScoreAcc <- try(read.table(paste(filename, ".scores.txt", sep =""), header = FALSE, as.is = TRUE))
  jsonstat_per <- try(fromJSON(paste(filename, ".scores.txt.matches.json", sep = "")))
  if(inherits(likeLiwind, 'try-error') | inherits(ScoreAcc, 'try-error') |  inherits(jsonstat_per, 'try-error')){
    next
  }
  
  plate <- c(plate, id)
  ranks <- order(-ScoreAcc$V4, na.last = TRUE, partial = ScoreAcc$V6)
  topscore <- ScoreAcc$V4[ranks[1]]
  topacc <- as.character(ScoreAcc$V1[ranks[1]])
  nextacc <- as.character(ScoreAcc$V1[ranks[2]])
  nextscore <- ScoreAcc$V4[ranks[2]]
  frac <- nextscore/topscore
  snps <- ScoreAcc$V7[ranks[1]]
  topnum <- length(which(ScoreAcc$V6 < 3.841))
  maxlike <- ScoreAcc$V5[ranks[1]]
  nextlike <- ScoreAcc$V6[ranks[2]]
  PerHeterozygosity = c(PerHeterozygosity, jsonstat_per$percent_heterozygosity)
  Overlap = c(Overlap, jsonstat_per$overlap[1])

  if (file.exists( paste(filename, ".matches.json", sep = "") )){
    jsonstat <- try(fromJSON(paste(filename, ".matches.json", sep = "")))
    FouParents1 <- c(FouParents1, as.character( jsonstat$parents$mother[1] ))
    FouParents2 <- c(FouParents2, as.character( jsonstat$parents$father[1] ))
    CountP1 <- c(CountP1, as.numeric( jsonstat$parents$mother[2] ))
    CountP2 <- c(CountP2, as.numeric( jsonstat$parents$father[2] ))
  } else {
    FouParents1 <- c(FouParents1, "")
    FouParents2 <- c(FouParents2, "")
    CountP1 <- c(CountP1, 0)
    CountP2 <- c(CountP2, 0)
  }

  ambWind <- which(likeLiwind$V7 < 20)
  nclean <- sort(-table(likeLiwind$V1[ambWind]))
  nclean <- as.data.frame(nclean)
  nclean <- nclean[which(nclean$nclean <= -1),]
  AmbiWindows <- c(AmbiWindows, paste(rownames(nclean), collapse = ":"))
  AmbiWindowCount <- c(AmbiWindowCount, paste(as.numeric(-nclean), collapse = ":"))
  NumWindows <- c(NumWindows, max(likeLiwind$V8))

  LikeLihoodTopHit <- c(LikeLihoodTopHit, maxlike)
  LLRNextHit <- c(LLRNextHit, nextlike)
  SNPscalled <- c(SNPscalled, snps)
  TopHit <- c(TopHit, topacc)
  NextHit <- c(NextHit, nextacc)
  TopScore <- c(TopScore, topscore)
  FracScore <- c(FracScore, frac)
  topnum <- length(which(ScoreAcc$V6 < 3.841))
  TopHitsNumber <- c(TopHitsNumber, topnum)
  Names <- c(Names, name)
  GivParents <- c(GivParents, expParents)
}

DF <- cbind(PLATE = plate, FILENAME = Names, ExpectedParents = GivParents, TopHit = TopHit, NextHit = NextHit, TopScore = TopScore, FracScore = FracScore, TopHitsNumber = TopHitsNumber, LikelihoodRatio = LikeLihoodTopHit, NextHitLLR = LLRNextHit, SNPsCalled = SNPscalled, ObservedParent1 = FouParents1, ObservedParent2 = FouParents2, CountP1 = CountP1, CountP2 = CountP2, ScoreP1 = ScoreP1, ScoreP2 = ScoreP2, Overlap = Overlap, Heterozygosity = PerHeterozygosity, AmbigousWindows = AmbiWindows, AmbigousWindowCount = AmbiWindowCount)

warnings()

write.csv(DF, file = outFile)

File: scripts/02_makeCSVTable_csmatch.py
# Generate a CSV file based on the ScoreAcc files generated before
# Input:
# 1. Output file
# 2. Folder ID that needs to be added
import numpy as np
import pandas as pd
import argparse
import logging
from glob import glob
from snpmatch.core import snpmatch
from snpmatch.core import snp_genotype
import os.path
import json
import re

logging.basicConfig(format='%(levelname)s:%(asctime)s: %(name)s:  %(message)s', level=logging.DEBUG)
log = logging.getLogger(__name__)

inOptions = argparse.ArgumentParser(description='get a genotype matrix')
inOptions.add_argument("-i", dest="files_path", default = "./", help="path to the snpmatch output files")
inOptions.add_argument("--file_spearation", dest="file_sep", default = ".", type = str, help="file name to file id (is it . separated or _)")
inOptions.add_argument("--dirs", action="store_true", dest="dirs", default=False, help="Input files are in folders")
inOptions.add_argument("-d", dest="accDB", help="hdf5 file for database accessions")
inOptions.add_argument("-e", dest="expParents", help="expected parents, ex: 6090x9416")
inOptions.add_argument("-f", dest="folder_id", help="Plate ID to be added in the CSV file")
inOptions.add_argument("-o", dest="output_file", default = "final_csmatch_genotyper", type=str, help="output file")

args = inOptions.parse_args()

log.info("reading input files")
if args.dirs:
    input_files = glob(args.files_path + "/csmatch_*/" + "*.windowscore.txt")
else:
    input_files = glob(args.files_path + "/" + "*.windowscore.txt")

input_files = pd.Series(input_files).astype(str)
assert len(input_files) > 0, "there are  no files with snpmatch scores in the folder."
input_ids = input_files.apply( os.path.basename ).str.split(args.file_sep, expand=True)
if np.unique(input_ids.iloc[:,0]).shape[0] == input_ids.shape[0]:
    input_ids = np.array(input_ids.iloc[:,0])
elif np.unique(input_ids.iloc[:,0] + args.file_sep + input_ids.iloc[:,1]).shape[0] == input_ids.shape[0]:
    input_ids = np.array(input_ids.iloc[:,0] + args.file_sep + input_ids.iloc[:,1])
else:
    input_ids = input_files.apply( os.path.basename ).str.replace( ".windowscore.txt", "" ).values

input_files = input_files.values

main_output_cols = ['FOL', "FILENAME", "ExpectedParents", 'TopHit', 'NextHit', 'Score', 'FracScore', 'SNPsCalled', 'LikelihoodRatio','NextHitLLR', 'TopHitsNumber', "ObservedParent1", "ObservedParent2", "CountP1", "CountP2", 'percent_heterozygosity', 'Overlap', "IdenticalWindows", "NumberofWindows", "HomozygousWindow", "HomozygousWindowCount"]
main_output = pd.DataFrame(columns = main_output_cols, index = input_ids)
main_output['FOL'] = args.folder_id
main_output['ExpectedParents'] = args.expParents

snp_db = snp_genotype.load_genotype_files(args.accDB)
sample_ancestry = pd.DataFrame(index = snp_db.accessions, columns =  input_ids)

assert input_ids.shape == np.array(input_files).shape, "Something is wrong"

for ef_ix in range(len(input_ids)):
    window_data = pd.read_csv(input_files[ef_ix],sep = "\t", dtype = {'acc': str})
    ScoreAcc = pd.read_csv(re.sub("windowscore.txt$", "scores.txt", input_files[ef_ix]) , header = None, sep = "\t")
    ScoreAcc = ScoreAcc.sort_values([5, 3], ascending=[True, False])
    main_output.loc[input_ids[ef_ix], "FILENAME"] = re.sub(".windowscore.txt$", "", input_files[ef_ix])
    main_output.loc[input_ids[ef_ix], "TopHit"] = ScoreAcc.iloc[0,0]
    main_output.loc[input_ids[ef_ix], "NextHit"] = ScoreAcc.iloc[1,0]
    main_output.loc[input_ids[ef_ix], "Score"] = ScoreAcc.iloc[0,3]
    main_output.loc[input_ids[ef_ix], "FracScore"] = ScoreAcc.iloc[1,3] / ScoreAcc.iloc[0,3]
    main_output.loc[input_ids[ef_ix], "SNPsCalled"] = ScoreAcc.iloc[0, 6]
    main_output.loc[input_ids[ef_ix], "LikelihoodRatio"] = ScoreAcc.iloc[0,4]
    main_output.loc[input_ids[ef_ix], "NextHitLLR"] = ScoreAcc.iloc[1,5]
    main_output.loc[input_ids[ef_ix], "TopHitsNumber"] = np.where(ScoreAcc.iloc[:,5] < snpmatch.lr_thres)[0].shape[0]
    num_winds = np.unique(window_data['window_index']).shape[0]
    main_output.loc[input_ids[ef_ix], "NumberofWindows"] = num_winds
    identical_wind = np.where(window_data.groupby('window_index').max()['identical'] == 1)[0].shape[0]
    sample_ancestry[input_ids[ef_ix]] = window_data.groupby('acc').mean()['identical'].reindex(sample_ancestry.index)
    main_output.loc[input_ids[ef_ix], "IdenticalWindows"] = snpmatch.get_fraction( identical_wind, num_winds )
    if not os.path.isfile(re.sub("windowscore.txt$", "matches.json", input_files[ef_ix])):
        continue
    with open(re.sub("windowscore.txt$", "matches.json", input_files[ef_ix])) as json_out:
        jsonstat = json.load(json_out)
    main_output.loc[input_ids[ef_ix], "ObservedParent1"] = jsonstat['parents']['mother'][0]
    main_output.loc[input_ids[ef_ix], "CountP1"] = jsonstat['parents']['mother'][1]
    main_output.loc[input_ids[ef_ix], "ObservedParent2"] = jsonstat['parents']['father'][0]
    main_output.loc[input_ids[ef_ix], "CountP2"] = jsonstat['parents']['father'][1]
    main_output.loc[input_ids[ef_ix], "percent_heterozygosity"] = jsonstat['percent_heterozygosity']
    main_output.loc[input_ids[ef_ix], "Overlap"] = jsonstat['overlap'][0]
    if len(jsonstat['matches']) > 0:
        main_output.loc[input_ids[ef_ix], "HomozygousWindow"] = pd.DataFrame(jsonstat['matches']).iloc[0:20,0].astype(str).str.cat(sep=',')
        main_output.loc[input_ids[ef_ix], "HomozygousWindowCount"] = pd.DataFrame(jsonstat['matches']).iloc[0:20,1].astype(str).str.cat(sep=',')
    if (ef_ix + 1) % 100 == 0:
        log.info("progress: %s files, %s%%" % (ef_ix, (ef_ix * 100 / len(input_ids)) ))


log.info("writing output into %s" % args.output_file )
main_output.to_csv(args.output_file + ".csv")
sample_ancestry.to_csv(args.output_file + "iden_ancestry.csv")
log.info("finished!")
